KOGITO-879 KOGITO-880 KOGITO-881: State Control API (undo/redo)

pefernan opened the pull request:
KOGITO-879: Define State Control API
KOGITO-880: Create JSInterop layer for StateControl API
KOGITO-881: Integrate State Control API in Stunner to handle Undo & Redo

On this PR I moved CommandRegistry into appformer level, so the injected registry will depend on the context (Business Central or Kogito). The command execution remains the same, only changes where the commands are being stored. 

@romartin @manstis would you mind take a look?

This is part of an ensemble, please merge with:
https://github.com/kiegroup/kogito-tooling/pull/66
https://github.com/kiegroup/droolsjbpm-build-bootstrap/pull/1195
https://github.com/kiegroup/appformer/pull/894
https://github.com/kiegroup/kie-wb-common/pull/3156
https://github.com/kiegroup/jbpm-wb/pull/1414
https://github.com/kiegroup/drools-wb/pull/1310
https://github.com/kiegroup/kie-wb-distributions/pull/1013

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/session/command/impl/RedoSessionCommand.java:
```diff
@@ -44,26 +45,34 @@
-        session.getKeyboardControl().addKeyShortcutCallback(this::onKeyDownEvent);
         redoCommandHandler.setSession(getSession());
+
+        //If running in Kogito we should initialize the Kogito StateControl undo/redo commands. Otherwise we should keep the key binding.
```
I struggle with this...

I'd much prefer `RedoSessionCommand` to be written for Business Central (i.e. binds to `session.getKeyboardControl()..` control, L#74). Then have another (CDI) `@Specializes` implementation [here](https://github.com/kiegroup/kie-wb-common/tree/master/kie-wb-common-kogito/kie-wb-common-kogito-webapp-base) that works with `KogitoStateControlInitializer`).

I'd also like to have seen more re-use of `EditorContextProvider.getChannel()` as opposed to `KogitoStateControlInitializer.isKogitoEnabled()` to determine the environment in which the editor is running.

manstis replied:
@romartin FYI... my suggestion would mean BPMN-kogito would need to extend the referenced module (IIUC that is yet to be done.. but this could be a good catalyst to ensure it's completed sooner than later!)

pefernan replied:
@manstis I think you're right on the @Specializes, Monday I'll work on that. The change that introduced the channel wasn't there when I started, I'll reuse that too.

pefernan commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/session/DMNEditorSession.java:
```diff
@@ -68,11 +70,13 @@
                             final SessionCommandManager<AbstractCanvasHandler> sessionCommandManager,
-                            final ClientCommandRegistry<org.kie.workbench.common.stunner.core.command.Command<AbstractCanvasHandler, CanvasViolation>> clientCommandRegistry) {
+                            final CommandRegistry<org.kie.workbench.common.stunner.core.command.Command<AbstractCanvasHandler, CanvasViolation>> commandRegistry,
+                            final Event<RegisterChangedEvent> registerChangedEvent) {
```
@manstis good question. The old ClientCommandRegistry was firing this event when changes on the registry were happening. Honestly I don't know why it's used, I just wrapped the event trigger on the RCL.

pefernan commented:
@manstis @jomarko could you please take a look again?

manstis commented on kie-wb-common-dmn/kie-wb-common-dmn-webapp-standalone/pom.xml:
```diff
@@ -537,6 +537,17 @@
+    <dependency>
+      <groupId>org.uberfire</groupId>
+      <artifactId>appformer-kogito-bridge</artifactId>
+      <scope>provided</scope>
```
I agree with @jomarko (and it's a great spot!) only the `-kogito` related modules should have dependencies on other `-kogito` modules which suggests a restructuring of the code in these PRs is needed (to move interfaces and _default_ - Business Central - implementations from `appformer-kogito-bridge` to some where more generic in `appformer` and keep kogito specific implementations alone in `appformer-kogito-bridge` etc).

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/test/java/org/kie/workbench/common/stunner/core/client/session/command/impl/UndoSessionCommandTest.java:
```diff
@@ -167,4 +167,28 @@ public void testOnCommandExecutedCheckWrongSession() {
         verify(commandRegistry,
                never()).clear();
     }
+/*
```
Did you want to remove this block?

pefernan commented:
@jomarko @manstis fixed the deps issues. I moved the basic APIs away from the bridge, then I moved the Kogito Undo/Redo Commands to a separate module to make them available to the editors runtime webapps. (They use the bridge to access the statecontrol APIs)

jomarko commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/session/command/impl/RedoSessionCommand.java:
```diff
@@ -62,8 +61,13 @@ public RedoSessionCommand(final SessionCommandManager<AbstractCanvasHandler> ses
-        session.getKeyboardControl().addKeyShortcutCallback(this::onKeyDownEvent);
         redoCommandHandler.setSession(getSession());
+
+        bindCommand();
```
I think should be added to `AbstractClientSessionCommand` (or higher) interface level with javadoc what is purpose of this method

pefernan replied:
The problem here is that those methods are only required for undo/redo sessioncommand, moving it up on the inheritance tree could cause a big refactor.

pefernan commented on kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-runtime/src/main/resources/org/kie/workbench/common/dmn/showcase/DMNKogitoRuntimeWebapp.gwt.xml:
```diff
@@ -19,6 +19,7 @@
 <module>
 
   <inherits name="org.kie.workbench.common.dmn.webapp.kogito.common.DMNWebappKogitoCommon"/>
+  <inherits name="org.kie.workbench.common.stunner.kogito.runtime.StunnerKogitoRuntimeCommon"/>
```
Actually the `StunnerKogitoRuntimeCommon` is already doing it.

pefernan commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/session/DMNEditorSessionTest.java:
```diff
@@ -20,6 +20,7 @@
 import java.util.Map;
 
 import com.google.gwtmockito.GwtMockitoTestRunner;
+import org.appformer.kogito.api.stateControl.registry.CommandRegistry;
```
It has a dep to `kie-wb-common-stunner-client-common` that already has a dep to `appformer-kogito-api`.

pefernan commented:
@manstis in the end the module is `appformer-client-api`

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-kogito/kie-wb-common-stunner-kogito-runtime-common/pom.xml:
```diff
@@ -0,0 +1,50 @@
+  <dependencies>
+    <dependency>
+      <groupId>org.uberfire</groupId>
+      <artifactId>appformer-kogito-bridge</artifactId>
```
scope shoudn't be `provided`?

pefernan replied:
@romartin I thought the provided was only relevant for showcase webaps, np with that.

The idea of `kie-wb-common-stunner-kogito-runtime-common` is to keep the base stunner stuff for kogito webaps (here I added the specific specialization of Undo/RedoSessionCommands for kogito). To me is OK moving stuff from `kie-wb-common-stunner-kogito-runtime-common` to `kie-wb-common-kogito-webapp-base` module, but this means that `kie-wb-common-kogito-webapp-base` will have deps to `kie-wb-common-stunner-client-common`. Not sure if that's something we want but if we all agree I'll do it.

manstis replied:
@romartin @pefernan `kie-wb-common-kogito-webapp-base` is where we've been keeping (client-side) stuff that is common to the webapps needed for `kogito` editors; so it makes sense to merge @pefernan new module with it. That said I'm happy with the PRs so any further improvements are "icing on the cake".

pefernan replied:
Ok, if you both agree going to move everything there, to me that's fine.

romartin replied:
so what I'm only trying to achieve is not adding new modules guys, because that makes everything complex...

On the other hand, as @manstis describes above, the `kie-wb-common-kogito-webapp-base` is a common module for `kogito` editors, which I understand is not just about Stunner, so I don't like the idea ok making `kie-wb-common-kogito-webapp-base` module depend on Stunner (`kie-wb-common-stunner-client-common`), but we don't have other quick solution I guess.

So well I will not block the PR for this too, was just a suggestion - given the small amount of code, no need for a new module, and also because we recently added some "common" module too... too many "common" :) But anyway, up to you guys as well, it's my own opinion but yours counts as mine!

romartin replied:
@pefernan good point about `provided` , my bad, it makes no difference here!

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-kogito/kie-wb-common-stunner-kogito-runtime-common/pom.xml:
```diff
@@ -0,0 +1,50 @@
+
+    <dependency>
+      <groupId>org.kie.workbench.stunner</groupId>
+      <artifactId>kie-wb-common-stunner-client-api</artifactId>
```
scope shoudn't be `provided`?

romartin replied:
nope, it should not :) Sorry

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-kogito/kie-wb-common-stunner-kogito-runtime-common/pom.xml:
```diff
@@ -0,0 +1,50 @@
+
+    <dependency>
+      <groupId>org.kie.workbench.stunner</groupId>
+      <artifactId>kie-wb-common-stunner-client-common</artifactId>
```
scope shoudn't be `provided`?

romartin replied:
nope, it should not :) Sorry

pefernan commented:
jenkins retest this please

pefernan commented:
@romartin review me!

tiagobento commented:
FDB passed. See https://github.com/kiegroup/appformer/pull/894
